### PR TITLE
Allow the use of magic methods to describe container methods

### PR DESCRIPTION
**Pull request by EmilyBourne**
Blocked by #2015

### CODE REVIEW

File: pyccel/naming/cnameclashchecker.py [hunk 1]
```diff
@@ -81,9 +81,10 @@ def get_collisionless_name(self, name, symbols):
         str
             A new name which is collision free.
         """
-        if len(name)>4 and all(name[i] == '_' for i in (0,1,-1,-2)):
-            # Ignore magic methods
+        if name in ('__init__', '__del__'):
             return name
+        if len(name)>4 and all(name[i] == '_' for i in (0,1,-1,-2)):
+            name = 'operator' + name[1:-2]
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/naming/cnameclashchecker.py:87]: /bot accept Renaming operators will be necessary when the wrapper support allows this to work for classes

File: pyccel/naming/fortrannameclashchecker.py [hunk 2]
```diff
@@ -91,9 +91,10 @@ def get_collisionless_name(self, name, symbols):
         str
             A new name which is collision free.
         """
-        if len(name)>4 and all(name[i] == '_' for i in (0,1,-1,-2)):
-            # Ignore magic methods
+        if name in ('__init__', '__del__'):
             return name
+        if len(name)>4 and all(name[i] == '_' for i in (0,1,-1,-2)):
+            name = 'operator' + name[1:-2]
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/naming/fortrannameclashchecker.py:97]: /bot accept Renaming operators will be necessary when the wrapper support allows this to work for classes

File: pyccel/parser/semantic.py [hunk 3]
```diff
@@ -905,11 +919,40 @@ def _create_PyccelOperator(self, expr, visited_args):
...
+        class_type = arg1.class_type
+        class_base = self.scope.find(str(class_type), 'classes') or get_cls_base(class_type)
+        magic_method_name = magic_method_map.get(type(expr), None)
+        magic_err = None
+        magic_method = None
+        if magic_method_name:
+            try:
+                magic_method = class_base.get_method(magic_method_name, False)
+            except PyccelSemanticError as err:
+                magic_err = err
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:932]: /bot accept
This line is just here to collect the error.

File: pyccel/parser/semantic.py [hunk 4]
```diff
@@ -905,11 +919,40 @@
...
+        if magic_err and len(visited_args) == 2:
+            arg2 = visited_args[1]
+            class_type = arg2.class_type
+            class_base = self.scope.find(str(class_type), 'classes') or get_cls_base(class_type)
+            magic_method_name = '__r'+magic_method_name[2:]
+            try:
+                magic_method = class_base.get_method(magic_method_name, False)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:939]: /bot accept
This code is almost identical to the code above so I am confident that it works but it is a corner case so I don't have something to test it with at the moment.

File: pyccel/parser/semantic.py [hunk 5]
```diff
@@ -905,11 +919,32 @@ def _create_PyccelOperator(self, expr, visited_args):
...
+                    visited_args = [visited_args[1], visited_args[0]]
+        if magic_method:
+            expr_new = self._handle_function(expr, magic_method, [FunctionCallArgument(v) for v in visited_args])
+        else:
+            try:
+                expr_new = type(expr)(*visited_args)
+            except PyccelSemanticError as err:
+                errors.report(str(err), symbol=expr, severity='fatal')
+            except TypeError as err:
+                types = ', '.join(str(a.class_type) for a in visited_args)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:945]: /bot accept
This is part of the error handling

File: pyccel/parser/semantic.py [hunk 6]
```diff
@@ -3376,6 +3420,52 @@
...
+            except PyccelSemanticError as err:
+                print(err)
+            operator_node = self._handle_function(expr, magic_method, [lhs, rhs])
+            lhs = self._assign_lhs_variable(expr.lhs, self._infer_type(operator_node), test_node,
+                    new_expressions, is_augassign = True)
+            lhs = self._optional_params.get(lhs, lhs)
+            aug_assign = Assign(lhs, operator_node)
+        if new_expressions:
+            return CodeBlock(new_expressions + [aug_assign])
+        else:
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:3453]: /bot accept
This is a generalisation, it is included for completeness

File: pyccel/ast/core.py [hunk 7]
```diff
@@ -3633,6 +3653,10 @@ def get_method(self, name):
         name : str
             The name of the attribute we are looking for.
 
+        raise_error : bool, default=True
+            True if an error should be found if the method is not found.
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/core.py:3657]:
```suggestion
            True if an error should be raised, False if None should be returned if the method is not found.
```

**Review by jalalium — APPROVED** [review]
Well done!
